Add new SqlDelight file action

romtsn opened the pull request:
Decided to start small :) most of the stuff is following how Kotlin plugin does it, also didn't really find a way how to test it (other plugins also don't really test actions), would be happy to receive some hints. Fix #42 

![new-file-template](https://user-images.githubusercontent.com/4999776/83658851-0f2c7900-a5c3-11ea-9f27-88e93e1f068f.gif)

romtsn commented on sqldelight-idea-plugin/src/main/resources/META-INF/plugin.xml:
```diff
@@ -22,14 +22,29 @@
   </project-components>
 
   <extensions defaultExtensionNs="com.intellij">
-    <fileTypeFactory implementation="com.squareup.sqldelight.intellij.lang.SqlDelightFileTypeFactory"/>
```
I've removed it as it's deprecated now, and we have to use `<fileType>` for that

AlecStrong replied:
:+1:

AlecStrong replied:
good callout. We can remedy that by changing the minimum version for the plugin, it'll prevent old versions of studio/intellij from downloading the plugin

romtsn replied:
hmm, should I squeeze this into the PR? I could also revert that actually, if you don't wanna break old idea clients

AlecStrong replied:
nah dont worry about it in this pr we can add it before releasing

romtsn commented on sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt:
```diff
@@ -0,0 +1,186 @@
+class SqlDelightCreateFileAction : CreateFileFromTemplateAction(
+    CAPTION,
+    "Creates new SqlDelight file or table",
+    SqlDelightFileType.icon
```
I didn't really find a way how to dynamically supply the icon for the action - looks like they are being created even before the projects is loaded (`ProjectManager.getInstance().openProjects` is empty by this time). So the icon in the menu will be always static it seems.. therefore I'm not sure whether I should use a static icon for the dialog as well or use the iconprovider

AlecStrong replied:
yup i think this is a reasonable default. If we want to make it unique or do something different later we can handle then

romtsn commented on sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt:
```diff
@@ -0,0 +1,186 @@
+    val propsFile = SqlDelightPropertiesFile.fromText(props.inputStream.reader().readText())!!
+
+    // only show the SqlDelight.New option if the selected folder is under src/**/sqldelight or a custom one
+    return propsFile.databases.first().compilationUnits
```
This means it's available only when you right-click on one of the `sourceFolders` or its subdirectories (`sqldelight` by default) - not sure if it's the best way, or should we show the action somewhere up in the hierarchy (e.g. `src`)?

AlecStrong replied:
this is really clever, and i think exactly the functionality we want. nice

AlecStrong commented on sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt:
```diff
@@ -0,0 +1,186 @@
+  override fun getActionName(directory: PsiDirectory?, newName: String, templateName: String?): String = CAPTION
+
+  override fun buildDialog(project: Project?, directory: PsiDirectory?, builder: CreateFileFromTemplateDialog.Builder) {
+    val icon = iconProvider.getIcon(dummyFile, 0, project)
```
i like using the iconProvider here, it's more accurate to the file that's about to be created and the icon it will have

AlecStrong replied:
i think we should add an api to `SqlDelightFileIconProvider` that returns the icon for a language/project, and then that way you don't need the dummyFile type

romtsn replied:
Good point, will do 👍

AlecStrong commented on sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt:
```diff
@@ -0,0 +1,186 @@
+    private fun createFileFromTemplate(name: String, template: FileTemplate, dir: PsiDirectory): PsiFile? {
+      val (className, targetDir) = findOrCreateTarget(dir, name, template.extension)
+
+      val service = DumbService.getInstance(dir.project)
```
were these just copied over? curious if not having it here causes any issues

romtsn replied:
Yep they were, but I'll check, I also did not really understand the purpose of those

AlecStrong commented:
one thing that would be sweet would be to verify migration files have a number somewhere in them, it looks like NameValidator works on all file types so that might not be possible. not blocking for this PR but curious if you thought that might be possible

romtsn commented:
Addressed the comments 👍  Re. the validator - I've checked and it basically means re-implementing the `CreateFileFromTemplateAction` and the dialog ourselves - I think it's a bit of overkill for such a hint :) so I went for somewhat ok-ish middle-ground - detect if there are words like `migration` or `sqm` in the entered name and report an error if there are no numbers. I can follow up in another PR, if you see it absolutely necessary

AlecStrong commented:
nope that works for me! thanks
